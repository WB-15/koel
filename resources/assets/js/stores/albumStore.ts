import { difference, orderBy, take, union } from 'lodash'

import { artistStore } from '.'
import { httpService } from '@/services'
import { arrayify, use } from '@/utils'
import { reactive } from 'vue'

interface AlbumStoreState {
  albums: Album[]
}

const UNKNOWN_ALBUM_ID = 1

export const albumStore = {
  cache: {} as Record<number, Album>,

  state: reactive<AlbumStoreState>({
    albums: []
  }),

  init (albums: Album[]) {
    // Traverse through the artists array and add their albums into our master album list.
    this.all = albums
    this.all.forEach(album => this.setupAlbum(album))
  },

  setupAlbum (album: Album) {
    const artist = artistStore.byId(album.artist_id)!
    artist.albums = union(artist.albums, [album])

    album.artist = artist
    album.info = null
    album.songs = []
    album.playCount = 0

    this.cache[album.id] = album
  },

  get all () {
    return this.state.albums
  },

  set all (value) {
    this.state.albums = value
  },

  byId (id: number) {
    return this.cache[id]
  },

  byIds (ids: number[]) {
    const albums = [] as Album[]
    ids.forEach(id => use(this.byId(id), album => albums.push(album!)))
    return albums
  },

  add (albums: Album | Album[]) {
    arrayify(albums).forEach(album => {
      this.setupAlbum(album)
      album.playCount = album.songs.reduce((count, song) => count + song.playCount, 0)
      this.all.push(album)
    })
  },

  /**
   * Remove empty albums from the store.
   */
  compact () {
    const emptyAlbums = this.all.filter(album => album.songs.length === 0)
    if (!emptyAlbums.length) {
      return
    }

    this.all = difference(this.all, emptyAlbums)
    emptyAlbums.forEach(album => delete this.cache[album.id])
  },

  getMostPlayed (count = 6): Album[] {
    // Only non-unknown albums with actual play count are applicable.
    const applicable = this.all.filter(album => album.playCount && !this.isUnknownAlbum(album))
    return take(orderBy(applicable, 'playCount', 'desc'), count)
  },

  getRecentlyAdded (count = 6): Album[] {
    const applicable = this.all.filter(album => !this.isUnknownAlbum(album))
    return take(orderBy(applicable, 'created_at', 'desc'), count)
  },

  /**
   * Upload a cover for an album.
   *
   * @param {Album} album The album object
   * @param {string} cover The content data string of the cover
   */
  uploadCover: async (album: Album, cover: string) => {
    album.cover = (await httpService.put<{ coverUrl: string }>(`album/${album.id}/cover`, { cover })).coverUrl
    return album.cover
  },

  /**
   * Get the (blurry) thumbnail-sized version of an album's cover.
   */
  getThumbnail: async (album: Album) => {
    if (album.thumbnail === undefined) {
      album.thumbnail = (await httpService.get<{ thumbnailUrl: string }>(`album/${album.id}/thumbnail`)).thumbnailUrl
    }

    return album.thumbnail
  },

  isUnknownAlbum: (album: Album) => album.id === UNKNOWN_ALBUM_ID
}
