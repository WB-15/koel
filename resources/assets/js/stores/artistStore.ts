import { difference, orderBy, take } from 'lodash'

import { httpService } from '@/services'
import stub from '@/stubs/artist'
import { arrayify, use } from '@/utils'
import { reactive } from 'vue'

interface ArtistStoreState {
  artists: Artist[]
}

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  stub,
  cache: {} as Record<number, Artist>,

  state: reactive<ArtistStoreState>({
    artists: []
  }),

  init (artists: Artist[]) {
    this.all = artists

    // Traverse through artists array to get the cover and number of songs for each.
    this.all.forEach(artist => this.setupArtist(artist))
  },

  setupArtist (artist: Artist) {
    artist.info = artist.info || null
    artist.albums = artist.albums || []
    artist.songs = artist.songs || []
    artist.playCount = artist.songs.reduce((count, song) => count + song.playCount, 0)

    this.cache[artist.id] = artist

    return artist
  },

  get all () {
    return this.state.artists
  },

  set all (value: Artist[]) {
    this.state.artists = value
  },

  byId (id: number): Artist | undefined {
    return this.cache[id]
  },

  byIds (ids: number[]) {
    const artists = [] as Artist[]
    ids.forEach(id => use(this.byId(id), artist => artists.push(artist!)))
    return artists
  },

  add (artists: Artist | Artist[]) {
    arrayify(artists).forEach(artist => this.all.push(this.setupArtist(artist)))
  },

  prepend (artists: Artist | Artist[]) {
    arrayify(artists).forEach(artist => this.all.unshift(this.setupArtist(artist)))
  },

  /**
   * Remove empty artists from the store.
   */
  compact () {
    const emptyArtists = this.all.filter(artist => artist.songs.length === 0)

    if (!emptyArtists.length) {
      return
    }

    this.all = difference(this.all, emptyArtists)
    emptyArtists.forEach(artist => delete this.cache[artist.id])
  },

  isVariousArtists: (artist: Artist) => artist.id === VARIOUS_ARTISTS_ID,

  isUnknownArtist: (artist: Artist) => artist.id === UNKNOWN_ARTIST_ID,

  getMostPlayed (count = 6): Artist[] {
    // Only non-unknown artists with actual play count are applicable.
    // Also, "Various Artists" doesn't count.
    const applicable = this.all.filter(artist => {
      return artist.playCount &&
        !this.isUnknownArtist(artist) &&
        !this.isVariousArtists(artist)
    })

    return take(orderBy(applicable, 'playCount', 'desc'), count)
  },

  /**
   * Upload an image for an artist.
   *
   * @param {Artist} artist The artist object
   * @param {string} image The content data string of the image
   */
  uploadImage: async (artist: Artist, image: string) => {
    const { imageUrl } = await httpService.put<{ imageUrl: string }>(`artist/${artist.id}/image`, { image })
    artist.image = imageUrl
    return artist.image
  }
}
