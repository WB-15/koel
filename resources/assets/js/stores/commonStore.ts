import isMobile from 'ismobilejs'
import { reactive } from 'vue'

import { httpService } from '@/services'
import {
  albumStore,
  artistStore,
  playlistStore,
  preferenceStore,
  queueStore,
  recentlyPlayedStore,
  settingStore,
  songStore,
  themeStore,
  userStore
} from '.'

interface CommonStoreState {
  albums: Album[]
  allowDownload: boolean
  artists: Artist[]
  cdnUrl: string
  currentUser: User | undefined
  currentVersion: string
  favorites: Song[]
  interactions: Interaction[]
  latestVersion: string
  originalMediaPath: string | undefined
  playlists: Playlist[]
  queued: Song[]
  recentlyPlayed: string[]
  settings: Settings
  songs: Song[]
  useiTunes: boolean
  useLastfm: boolean
  users: User[]
  useYouTube: boolean
}

export const commonStore = {
  state: reactive<CommonStoreState>({
    albums: [],
    allowDownload: false,
    artists: [],
    cdnUrl: '',
    currentUser: undefined,
    currentVersion: '',
    favorites: [],
    interactions: [],
    latestVersion: '',
    originalMediaPath: '',
    playlists: [],
    queued: [],
    recentlyPlayed: [],
    settings: {} as Settings,
    songs: [],
    useiTunes: false,
    useLastfm: false,
    users: [],
    useYouTube: false
  }),

  async init () {
    Object.assign(this.state, await httpService.get<CommonStoreState>('data'))

    // Always disable YouTube integration on mobile.
    this.state.useYouTube = this.state.useYouTube && !isMobile.phone

    // If this is a new user, initialize his preferences to be an empty object.
    this.state.currentUser!.preferences = this.state.currentUser!.preferences || {}

    userStore.init(this.state.users, this.state.currentUser!)
    preferenceStore.init(this.state.currentUser)
    artistStore.init(this.state.artists)
    albumStore.init(this.state.albums)
    songStore.init(this.state.songs)
    songStore.initInteractions(this.state.interactions)
    recentlyPlayedStore.initExcerpt(this.state.recentlyPlayed)
    playlistStore.init(this.state.playlists)
    queueStore.init()
    settingStore.init(this.state.settings)
    themeStore.init()

    return this.state
  }
}
