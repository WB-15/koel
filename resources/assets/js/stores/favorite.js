import _ from 'lodash';

import http from '../services/http';

export default {
    state: {
        songs: [],
        length: 0,
        fmtLength: '',
    },

    /**
     * All songs favorite'd by the current user.
     *
     * @return {Array.<Object>}
     */
    get all() {
        return this.state.songs;
    },

    clear() {
        this.state.songs = [];
    },

    /**
     * Toggle like/unlike a song.
     * A request to the server will be made.
     *
     * @param {Object}     song
     * @param {?Function}  cb
     */
    toggleOne(song, cb = null) {
        // Don't wait for the HTTP response to update the status, just toggle right away.
        // This may cause a minor problem if the request fails somehow, but do we care?
        song.liked = !song.liked;

        if (song.liked) {
            this.add(song);
        } else {
            this.remove(song);
        }

        http.post('interaction/like', { song: song.id }, () => {
            if (cb) {
                cb();
            }
        });
    },

    /**
     * Add a song into the store.
     *
     * @param {Object} song
     */
    add(song) {
        this.state.songs.push(song);
    },

    /**
     * Remove a song from the store.
     *
     * @param {Object} song
     */
    remove(song) {
        this.state.songs = _.difference(this.state.songs, [song]);
    },

    /**
     * Like a bunch of songs.
     *
     * @param {Array.<Object>}  songs
     * @param {?Function}       cb
     */
    like(songs, cb = null) {
        // Don't wait for the HTTP response to update the status, just set them to Liked right away.
        // This may cause a minor problem if the request fails somehow, but do we care?
        _.each(songs, song => song.liked = true);
        this.state.songs = _.union(this.state.songs, songs);

        http.post('interaction/batch/like', { songs: _.pluck(songs, 'id') }, () => {
            if (cb) {
                cb();
            }
        });
    },

    /**
     * Unlike a bunch of songs.
     *
     * @param {Array.<Object>}  songs
     * @param {?Function}       cb
     */
    unlike(songs, cb = null) {
        // Don't wait for the HTTP response to update the status, just set them to Unliked right away.
        // This may cause a minor problem if the request fails somehow, but do we care?
        _.each(songs, song => song.liked = false);
        this.state.songs = _.difference(this.state.songs, songs);

        http.post('interaction/batch/unlike', { songs: _.pluck(songs, 'id') }, () => {
            if (cb) {
                cb();
            }
        });
    },
};
