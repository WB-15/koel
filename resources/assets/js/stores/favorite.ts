import { difference, union } from 'lodash'
import { http } from '@/services'

export const favoriteStore = {
  state: {
    songs: [] as Song[],
    length: 0,
    fmtLength: ''
  },

  get all (): Song[] {
    return this.state.songs
  },

  set all (value: Song[]) {
    this.state.songs = value
  },

  async toggleOne (song: Song): Promise<void> {
    // Don't wait for the HTTP response to update the status, just toggle right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    song.liked = !song.liked
    song.liked ? this.add(song) : this.remove(song)

    await http.post<Song>('interaction/like', { song: song.id })
  },

  /**
   * Add a song/songs into the store.
   */
  add (songs: Song | Song[]): void {
    this.all = union(this.all, (<Song[]>[]).concat(songs))
  },

  /**
   * Remove a song/songs from the store.
   */
  remove (songs: Song | Song[]): void {
    this.all = difference(this.all, (<Song[]>[]).concat(songs))
  },

  clear (): void {
    this.all = []
  },

  async like (songs: Song[]): Promise<void> {
    // Don't wait for the HTTP response to update the status, just set them to Liked right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    songs.forEach(song => { song.liked = true })
    this.add(songs)

    await http.post('interaction/batch/like', { songs: songs.map(song => song.id) })
  },

  async unlike (songs: Song[]): Promise<void> {
    songs.forEach(song => { song.liked = false })
    this.remove(songs)

    await http.post('interaction/batch/unlike', { songs: songs.map(song => song.id) })
  }
}
