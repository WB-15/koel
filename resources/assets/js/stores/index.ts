export * from './albumStore'
export * from './artistStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './playlistStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './settingStore'
export * from './commonStore'
export * from './songStore'
export * from './userStore'
export * from './searchStore'
export * from './themeStore'
