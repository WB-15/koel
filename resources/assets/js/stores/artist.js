import Vue from 'vue';
import _ from 'lodash';

import config from '../config';
import stub from '../stubs/artist';
import albumStore from './album';

export default {
    stub,

    state: {
        artists: [],
    },

    /**
     * Init the store.
     *
     * @param  {Array.<Object>} artists The array of artists we got from the server.
     */
    init(artists) {
        this.state.artists = artists;

        // Traverse through artists array to get the cover and number of songs for each.
        _.each(this.state.artists, artist => {
            this.setupArtist(artist);
        });

        albumStore.init(this.state.artists);
    },

    setupArtist(artist) {
        this.getImage(artist);
        Vue.set(artist, 'playCount', 0);
        Vue.set(artist, 'songCount', _.reduce(artist.albums, (count, album) => count + album.songs.length, 0));
        Vue.set(artist, 'info', null);

        return artist;
    },

    get all() {
        return this.state.artists;
    },

    byId(id) {
        return _.find(this.all, 'id', id);
    },

    /**
     * Appends a new artist into the current collection.
     *
     * @param  {Object} artist
     */
    append(artist) {
        this.state.artists.push(this.setupArtist(artist));
    },

    addAlbumsIntoArtist(artist, albums) {
        albums = [].concat(albums);

        artist.albums = _.union(artist.albums ? artist.albums : [], albums);

        albums.forEach(album => {
            album.artist_id = artist.id;
            album.artist = artist;
        });

        artist.playCount = _.reduce(artist.albums, (count, album) => count + album.playCount, 0);
    },

    /**
     * Remove album(s) from an artist.
     *
     * @param  {Object} artist
     * @param  {Array.<Object>|Object} albums
     */
    removeAlbumsFromArtist(artist, albums) {
        artist.albums = _.difference(artist.albums, [].concat(albums));
        artist.playCount = _.reduce(artist.albums, (count, album) => count + album.playCount, 0);
    },

    /**
     * Checks if an artist is empty.
     *
     * @param  {Object}  artist
     *
     * @return {boolean}
     */
    isArtistEmpty(artist) {
        return !artist.albums.length;
    },

    /**
     * Remove artist(s) from the store.
     *
     * @param  {Array.<Object>|Object} artists
     */
    remove(artists) {
        this.state.artists = _.difference(this.state.artists, [].concat(artists));
    },

    /**
     * Get all songs performed by an artist.
     *
     * @param {Object} artist
     *
     * @return {Array.<Object>}
     */
    getSongsByArtist(artist) {
        if (!artist.songs) {
            artist.songs = _.reduce(artist.albums, (songs, album) => songs.concat(album.songs), []);
        }

        return artist.songs;
    },

    /**
     * Get the artist's image.
     *
     * @param {Object} artist
     *
     * @return {String}
     */
    getImage(artist) {
        // If the artist already has a proper image, just return it.
        if (artist.image) {
            return artist.image;
        }

        // Otherwise, we try to get an image from one of their albums.
        artist.image = config.unknownCover;

        artist.albums.every(album => {
            // If there's a "real" cover, use it.
            if (album.image !== config.unknownCover) {
                artist.image = album.cover;

                // I want to break free.
                return false;
            }
        });

        return artist.image;
    },

    /**
     * Get top n most-played artists.
     *
     * @param  {Number} n
     *
     * @return {Array.<Object>}
     */
    getMostPlayed(n = 6) {
        // Only non-unknown artists with actually play count are applicable.
        const applicable = _.filter(this.state.artists, artist => {
            return artist.playCount && artist.id !== 1;
        });

        return _.take(_.sortByOrder(applicable, 'playCount', 'desc'), n);
    },
};
