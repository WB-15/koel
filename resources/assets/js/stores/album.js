import Vue from 'vue';
import _ from 'lodash';

import utils from '../services/utils';
import stub from '../stubs/album';
import songStore from './song';
import artistStore from './artist';

export default {
    stub,

    state: {
        albums: [stub],
    },

    /**
     * Init the store.
     *
     * @param  {Array.<Object>} artists The array of artists to extract album data from.
     */
    init(artists) {
        // Traverse through the artists array and add their albums into our master album list.
        this.state.albums = _.reduce(artists, (albums, artist) => {
            // While we're doing so, for each album, we get its length
            // and keep a back reference to the artist too.
            _.each(artist.albums, album => {
                this.setupAlbum(album, artist);
            });

            return albums.concat(artist.albums);
        }, []);

        // Then we init the song store.
        songStore.init(this.state.albums);
    },

    setupAlbum(album, artist) {
        Vue.set(album, 'playCount', 0);
        Vue.set(album, 'artist', artist);
        Vue.set(album, 'info', null);
        this.getLength(album);

        return album;
    },

    /**
     * Get all albums in the store.
     *
     * @return {Array.<Object>}
     */
    get all() {
        return this.state.albums;
    },

    byId(id) {
        return _.find(this.all, 'id', id);
    },

    /**
     * Get the total length of an album by summing up its songs' duration.
     * The length will also be converted into a H:i:s format and stored as fmtLength.
     *
     * @param  {Object} album
     *
     * @return {String} The H:i:s format of the album length.
     */
    getLength(album) {
        Vue.set(album, 'length', _.reduce(album.songs, (length, song) => length + song.length, 0));
        Vue.set(album, 'fmtLength', utils.secondsToHis(album.length));

        return album.fmtLength;
    },

    /**
     * Appends a new album into the current collection.
     *
     * @param  {Object} album
     */
    append(album) {
        this.state.albums.push(this.setupAlbum(album, album.artist));
        album.playCount = _.reduce(album.songs, (count, song) => count + song.playCount, 0);
    },

    /**
     * Add song(s) into an album.
     *
     * @param {Object} album
     * @param {Array.<Object>|Object} song
     */
    addSongsIntoAlbum(album, songs) {
        songs = [].concat(songs);

        album.songs = _.union(album.songs ? album.songs : [], songs);

        songs.forEach(song => {
            song.album_id = album.id;
            song.album = album;
        });

        album.playCount = _.reduce(album.songs, (count, song) => count + song.playCount, 0);
        this.getLength(album);
    },

    /**
     * Remove song(s) from an album.
     *
     * @param  {Object} album
     * @param  {Array.<Object>|Object} songs
     */
    removeSongsFromAlbum(album, songs) {
        album.songs = _.difference(album.songs, [].concat(songs));
        album.playCount = _.reduce(album.songs, (count, song) => count + song.playCount, 0);
        this.getLength(album);
    },

    /**
     * Checks if an album is empty.
     *
     * @param  {Object}  album
     *
     * @return {boolean}
     */
    isAlbumEmpty(album) {
        return !album.songs.length;
    },

    /**
     * Remove album(s) from the store.
     *
     * @param  {Array.<Object>|Object} albums
     */
    remove(albums) {
        albums = [].concat(albums);
        this.state.albums = _.difference(this.state.albums, albums);

        // Remove from the artist as well
        albums.forEach(album => {
            artistStore.removeAlbumsFromArtist(album.artist, album);
        });
    },

    /**
     * Get top n most-played albums.
     *
     * @param  {Number} n
     *
     * @return {Array.<Object>}
     */
    getMostPlayed(n = 6) {
        // Only non-unknown albums with actually play count are applicable.
        const applicable = _.filter(this.state.albums, album => {
            return album.playCount && album.id !== 1;
        });

        return _.take(_.sortByOrder(applicable, 'playCount', 'desc'), n);
    },
};
