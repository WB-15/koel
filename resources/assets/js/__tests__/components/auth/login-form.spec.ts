import Component from '@/components/auth/LoginForm.vue'
import { userStore } from '@/stores'
import { mock } from '@/__tests__/__helpers__'
import { shallow } from '@/__tests__/adapter'

describe('components/auth/LoginForm', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders properly', () => {
    expect(shallow(Component)).toMatchSnapshot()
  })

  it('triggers login when form is submitted', () => {
    const loginStub = mock(userStore, 'login')
    shallow(Component, {
      data: () => ({
        email: 'john@doe.com',
        password: 'secret'
      })
    }).submit('form')
    expect(loginStub).toHaveBeenCalledWith('john@doe.com', 'secret')
  })
})
