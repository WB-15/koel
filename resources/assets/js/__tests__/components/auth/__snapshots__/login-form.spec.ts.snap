// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/auth/LoginForm renders properly 1`] = `
<form data-testid="login-form" class="">
  <div class="logo"><img src="@/../img/logo.svg" width="156" height="auto" alt="Koel's logo"></div>
  <!----> <input type="email" placeholder="Email Address" autofocus="autofocus" required="required"> <input type="password" placeholder="Password" required="required">
  <btn-stub type="submit">Log In</btn-stub>
</form>
`;
